### PR TITLE
Follow conventions for CMAKE_BUILD_TYPE

### PR DESCRIPTION
**Pull request by ktf**
Follow conventions for CMAKE_BUILD_TYPE

The way case sensitiveness works in CMake is described at:

https://cmake.org/cmake/help/v3.29/manual/cmake-buildsystem.7.html#case-sensitivity

and the "strong" convention is described a few lines below:

https://cmake.org/cmake/help/v3.29/manual/cmake-buildsystem.7.html#id38

Notice also how the variables for customising CMAKE_CXX_FLAGS (e.g. CMAKE_CXX_FLAGS_DEBUG)
are always all uppercase, regardless of the value of CMAKE_BUILD_TYPE. We use the
camel cased version, because that's what strongly suggested by CMake (probably because the
generators for some IDE / tool use that convention).

Moreover MinSizeRel is the typical spelling, not MinRelSize.

### CODE REVIEW

**Comment by davidrohr** [comment]
I think using the `$<CONFIG:...>` makes sense in any case.
However, the problem we had yesterday was that https://github.com/AliceO2Group/AliceO2/blob/20b04be86307a95ac0fd5bb09e849bb073bc1731/dependencies/O2CompileFlags.cmake#L145 was not working.
The `-ffp-contract=off` got forwarded correctly if CMAKE_BUILD_TYPE=RELWITHDEBINFO, but it failed with CMAKE_BUILD_TYPE=RelWithDebInfo. To be honest, I did not manage to figure out why it failed, but thought it would be safe to enforce UPPERCASE.
I am fine with and in favor of your solution if we make the appending do `${CMAKE_CXX_FLAGS_${CMAKE_BUILD_TYPE}}` (and alike) work. However, if we do so, should we also change alidist/defaults-o2.sh build type to RelWithDebInfo to be consistent? That would require to change plenty of other packages perhaps. Perhaps we should discuss in person...

**Comment by ktf** [comment]
@davidrohr @mconcas I think I took everything into account, as discussed. Closing the other one.

File: GPU/GPUTracking/Standalone/CMakeLists.txt [hunk 1]
```diff
@@ -50,7 +50,7 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 
 if(BUILD_DEBUG)
   set(CMAKE_CXX_FLAGS "-O0 -ggdb") # -fsanitize=address,undefined -fno-sanitize=vptr
-  set(CMAKE_BUILD_TYPE DEBUG)
+  set(CMAKE_BUILD_TYPE Debug)
```
Comment (davidrohr) [inline GPU/GPUTracking/Standalone/CMakeLists.txt]: Could you please not change the settings for the standalone build. Anyway not part of O2. And the alidist build type is also DEBUG / RELEASE, so no reason to enforce camel-case.

File: GPU/GPUTracking/Standalone/CMakeLists.txt [hunk 2]
```diff
@@ -61,7 +61,7 @@ else()
   if (NOT CMAKE_CXX_COMPILER STREQUAL "clang++")
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ftracer -fprefetch-loop-arrays")
   endif()
-  set(CMAKE_BUILD_TYPE RELEASE)
+  set(CMAKE_BUILD_TYPE Release)
```
Comment (davidrohr) [inline GPU/GPUTracking/Standalone/CMakeLists.txt]: same here

**Review by davidrohr — CHANGES_REQUESTED** [review]
Looks good to me except for 2 unnecessary changes below

**Review by davidrohr — CHANGES_REQUESTED** [review]
And actually, it does not compile like this:
```
  Syntax error in cmake code at

    /home/qon/alice/O2/dependencies/FindO2GPU.cmake:117

  when parsing string

    ${CMAKE_CUDA_FLAGS_$<UPPER_CASE:${CMAKE_BUILD_TYPE}>} -Xcompiler \"${CMAKE_CXX_FLAGS_$<UPPER_CASE:${CMAKE_BUILD_TYPE}>}\" -Xptxas -O4 -Xcompiler -O4

  Invalid character ('<') in a variable name: 'CMAKE_CUDA_FLAGS_$'
```

File: GPU/GPUTracking/Standalone/CMakeLists.txt (hunk 1 above)
Comment (ktf) [reply GPU/GPUTracking/Standalone/CMakeLists.txt]: Sure.

**Comment by ktf** [comment]
Indeed one needs to use normal variables, not generator expressions. Nethertheless I would not change CMAKE_BUILD_TYPE to be always uppercase and introduce a separate one, so that if the generator itself assumes Debug rather than DEBUG we are fine.

**Review by davidrohr — APPROVED** [review]

**Comment by ktf** [comment]
Ok, I understand what is the rationale, now. The idea is that ninja (and other generators) support a multi-config setup (since CMake 3.17), allowing different build types to coexists. This means that one could have (in the same build area) a "stage-Release" area and a "stage-Debug" area, where the byproducts of the two different build types are compiled (each with the respective options) and stored. One can then seamlessly mix and match them. In order for that to work, global compile options need to be added via something like:

```cmake
 add_compile_options($<$<CONFIG:Debug>:-O0> $<$<CONFIG:Debug>:-ggdb>)
 add_compile_options($<$<CONFIG:Release>:-O3> $<$<CONFIG:Release>:-ffast-math>)
```

rather than the current CMAKE_CXX_FLAGS setup. One of the advantages would be to have debug executables available in the same environment as the non debug ones (we could even add a -debug / -asan suffix to ease deployment).

If the above mode is used, the `CMAKE_BUILD_TYPE` is actually not set.

IMHO, it is probably worth investigating, even for things like GPU / non-GPU builds or hip vs CUDA, but it requires

**Comment by davidrohr** [comment]
Fair enough, I can change all my additions to that scheme. Though, I want to have all builds (debug and release) with GPUCA_NO_FAST_MATH flags, if that flag is said. So I'd need an option to set it for all build types, in a way that it overrides particular settings for a specific build type).

And then, How can I set it language-dependant, e.g. only for CUDA?

**Comment by davidrohr** [comment]
hm, ok, there is ` $<COMPILE_LANGUAGE>` for that. Still need to find out how to override build-type specific settings

**Comment by davidrohr** [comment]
errors unrelated, merging, we can improve upon this in another PR

**Comment by ktf** [comment]
> So I'd need an option to set it for all build types, in a way that it overrides particular settings for a specific build type).

I think that you do "old style":

```
if (GPUCA_NO_FAST_MATH)
add_compile_option(-fno-fast-math)
endif()
```

The above should apply to all targets. I think that's nevertheless the "new" way of adding global options.

**Pull request merged**